lib/ruby_lsp/requests/document_link.rb

```diff
@@ -1,30 +1,79 @@
 # typed: strict
 # frozen_string_literal: true
 
+require "ruby_lsp/requests/support/source_uri"
+
 module RubyLsp
   module Requests
     # ![Document link demo](../../misc/document_link.gif)
     #
     # The [document link](https://microsoft.github.io/language-server-protocol/specification#textDocument_documentLink)
-    # makes `# source://PATH_TO_FILE:line` comments in a Ruby/RBI file clickable if the file exists.
+    # makes `# source://PATH_TO_FILE#line` comments in a Ruby/RBI file clickable if the file exists.
     # When the user clicks the link, it'll open that location.
     #
     # # Example
     #
     # ```ruby
-    # # source://syntax_tree-3.2.1/lib/syntax_tree.rb:51 <- it will be clickable and will take the user to that location
+    # # source://syntax_tree/3.2.1/lib/syntax_tree.rb#51 <- it will be clickable and will take the user to that location
     # def format(source, maxwidth = T.unsafe(nil))
     # end
     # ```
     class DocumentLink < BaseRequest
       extend T::Sig
 
-      RUBY_ROOT = "RUBY_ROOT"
+      GEM_TO_VERSION_MAP = T.let(
+        [*::Gem::Specification.default_stubs, *::Gem::Specification.stubs].map! do |s|
+          [s.name, s.version.to_s]
+        end.to_h.freeze,
+        T::Hash[String, String]
+      )
+
+      class << self
+        extend T::Sig
+
+        sig { returns(T::Hash[String, T::Array[String]]) }
+        def gem_paths
+          @gem_paths ||= T.let(begin
+            lookup = {}
+
+            Gem::Specification.stubs.each do |stub|
+              spec = stub.to_spec
+              lookup[spec.name] = {}
+              lookup[spec.name][spec.version.to_s] = {}
+
+              Dir.glob("**/*.rb", base: "#{spec.full_gem_path}/").each do |path|
+                lookup[spec.name][spec.version.to_s][path] = "#{spec.full_gem_path}/#{path}"
+              end
+            end
+
+            Gem::Specification.default_stubs.each do |stub|
+              spec = stub.to_spec
+              lookup[spec.name] = {}
+              lookup[spec.name][spec.version.to_s] = {}
+
+              spec.files.each do |file|
+                prefix_matchers = [//]
+                prefix_matchers.concat(spec.require_paths.map { |rp| Regexp.new("^#{rp}/") })
+                prefix_matcher = Regexp.union(prefix_matchers)
```

Comment:
Since these 3 lines don't need access to `file`, maybe they can be moved out of the loop?